packages/core/core/src/RequestTracker.js

```diff
@@ -402,78 +420,93 @@ export class RequestGraph extends Graph<
       // This indicates a complete match, and any requests attached to the `aboveFilePath`
       // node will be invalidated.
       let firstId = 'file_name:' + parts[0];
-      if (!this.hasEdge(node.id, firstId, 'invalidated_by_create_above')) {
-        this.addEdge(node.id, firstId, 'invalidated_by_create_above');
+      let firstNodeId = this.getNodeIdByContentKey(firstId);
+      if (!this.hasEdge(nodeId, firstNodeId, 'invalidated_by_create_above')) {
+        this.addEdge(nodeId, firstNodeId, 'invalidated_by_create_above');
       }
 
-      invariant(last != null);
-      if (!this.hasEdge(last.id, node.id, 'invalidated_by_create_above')) {
-        this.addEdge(last.id, node.id, 'invalidated_by_create_above');
+      invariant(lastNodeId != null);
+      if (!this.hasEdge(lastNodeId, nodeId, 'invalidated_by_create_above')) {
+        this.addEdge(lastNodeId, nodeId, 'invalidated_by_create_above');
       }
     } else if (input.filePath != null) {
       node = nodeFromFilePath(input.filePath);
     } else {
       throw new Error('Invalid invalidation');
     }
 
-    if (!this.hasNode(node.id)) {
-      this.addNode(node);
+    let nodeId;
+    if (this.hasContentKey(node.id)) {
+      nodeId = this.getNodeIdByContentKey(node.id);
+    } else {
+      nodeId = this.addNode(node);
     }
```

Comment:
There are many occurrences of exact this pattern in this file, maybe this could be `graph.ensureNode(node: TNode): NodeID`

Reply:
Actually, should this rather be included in `addNode`? At the moment you could call `addNode(foo);` twice and the node will actually be inserted twice